alvr/server_core/src/connection.rs

```diff
@@ -265,16 +265,69 @@ pub fn handshake_loop(ctx: Arc<ConnectionContext>, lifecycle_state: Arc<RwLock<L
             manual_client_ips
         };
 
-        if !available_manual_client_ips.is_empty()
-            && try_connect(
+        if !available_manual_client_ips.is_empty() {
+            if available_manual_client_ips.keys().any(|i| i.is_loopback()) {
+                let adb_path = match alvr_adb::get_adb_path() {
+                    Some(path) => path,
+                    None => {
+                        dbg_connection!("Couldn't find adb, installing it...");
+                        if let Err(e) = alvr_adb::install_adb(|downloaded, total| {
+                            dbg_connection!("Downloaded {downloaded:?} bytes of {total:?}");
+                        }) {
+                            error!("Failed to install adb: {e:?}");
+                            thread::sleep(RETRY_CONNECT_MIN_INTERVAL);
+                            continue;
+                        }
+                        dbg_connection!("Finished installing adb");
+                        let maybe_adb_path = alvr_adb::get_adb_path();
+                        let Some(adb_path) = maybe_adb_path else {
+                            error!("Failed to get adb path after installation");
+                            thread::sleep(RETRY_CONNECT_MIN_INTERVAL);
+                            continue;
+                        };
+                        adb_path
+                    }
+                };
+                dbg_connection!("Found adb executable: {adb_path:?}");
+                let devices = match alvr_adb::list_devices(&adb_path) {
+                    Err(e) => {
+                        error!("Failed to list adb devices: {e:?}");
+                        thread::sleep(RETRY_CONNECT_MIN_INTERVAL);
+                        continue;
+                    }
+                    Ok(devices) => devices.into_iter().filter(|d| {
+                        d.serial
+                            .as_ref()
+                            .is_some_and(|s| !s.starts_with("127.0.0.1"))
+                    }),
+                };
+                let ports = HashSet::from([9943, 9944]);
+                for device in devices {
+                    let Some(device_serial) = device.serial else {
+                        dbg_connection!("Skipping device without serial number");
+                        continue;
```

Comment:
Sorry, maybe I was not clear, but the I think this whole section can be moved to alvr_adb. Because there are only references to alvr_adb functions